File: mvp/roguelike_survival_/src/systems/particles.py
import pygame
import random
import math

class Particle:
    __slots__ = ['x', 'y', 'color', 'vx', 'vy', 'size', 'max_life', 'life']
    def __init__(self, x, y, color, velocity, size, life):
        self.x = x
        self.y = y
        self.color = color
        self.vx = velocity[0]
        self.vy = velocity[1]
        self.size = size
        self.max_life = life
        self.life = life

    def update(self, dt):
        self.x += self.vx * dt
        self.y += self.vy * dt
        self.life -= dt

class ParticleSystem:
    def __init__(self):
        self.particles = []
        self._surface_cache = {}

    def emit(self, x, y, color, count=10, speed=100):
        for _ in range(count):
            angle = random.uniform(0, math.pi * 2)
            sp = random.uniform(speed * 0.5, speed * 1.5)
            vx = math.cos(angle) * sp
            vy = math.sin(angle) * sp
            size = random.uniform(2, 6)
            life = random.uniform(0.3, 0.9)
            self.particles.append(Particle(x, y, color, (vx, vy), size, life))

    def update(self, dt):
        alive = []
        for p in self.particles:
            p.update(dt)
            if p.life > 0:
                alive.append(p)
        self.particles = alive

    def _get_particle_surf(self, color, size, alpha):
        q_size = int(size)
        q_alpha = (int(alpha) // 10) * 10
        key = (color, q_size, q_alpha)
        
        if key not in self._surface_cache:
            if q_size <= 0: return None
            s = pygame.Surface((q_size * 2, q_size * 2), pygame.SRCALPHA).convert_alpha()
            pygame.draw.circle(s, (*color, q_alpha), (q_size, q_size), q_size)
            self._surface_cache[key] = s
        return self._surface_cache[key]

    def draw(self, surface, camera):
        for p in self.particles:
            alpha = int(255 * (p.life / p.max_life))
            size = max(1, p.size * (p.life / p.max_life))
            scaled_size = camera.apply_scale(size)
            
            p_surf = self._get_particle_surf(p.color, scaled_size, alpha)
            if p_surf:
                pos = camera.apply(p.x, p.y)
                surface.blit(p_surf, (pos[0] - scaled_size, pos[1] - scaled_size), special_flags=pygame.BLEND_ADD)

File: mvp/commander_sim/js/view/Particles.js
export class Particle {
    constructor(x, y, color, speed, life, type = 'spark', text = null) {
        this.x = x;
        this.y = y;
        this.color = color;
        this.life = life;
        this.maxLife = life;
        this.type = type; // spark, shockwave, text
        this.text = text;
        
        // Physics
        const angle = Math.random() * Math.PI * 2;
        // Shockwaves don't move, they expand
        const moveSpeed = type === 'shockwave' ? 0 : speed;
        
        this.vx = Math.cos(angle) * moveSpeed * (Math.random() + 0.5);
        this.vy = Math.sin(angle) * moveSpeed * (Math.random() + 0.5);
        
        if (type === 'text') {
            this.vx = (Math.random() - 0.5) * 1;
            this.vy = -2; // Float up
        }

        this.size = Math.random() * 3 + 2;
        if (type === 'shockwave') this.size = 5; // Start small
        
        this.alpha = 1;
    }

    update() {
        this.x += this.vx;
        this.y += this.vy;
        this.life--;
        
        // Normalize alpha
        this.alpha = Math.max(0, this.life / this.maxLife);

        if (this.type === 'spark') {
            this.vx *= 0.92; // Friction
            this.vy *= 0.92;
            this.size *= 0.96; // Shrink
        } else if (this.type === 'shockwave') {
            this.size += 3; // Expand rapidly
            this.alpha = Math.pow(this.alpha, 2); // Fade out faster visually
        } else if (this.type === 'text') {
            this.vy *= 0.98; // Slow rise
        }
    }
}

export class ParticleSystem {
    constructor() {
        this.particles = [];
        this.maxParticles = 800;
    }

    emit(x, y, color, count = 5, type = 'spark') {
        if (this.particles.length > this.maxParticles) return;
        for (let i = 0; i < count; i++) {
            this.particles.push(new Particle(x, y, color, 4, 20 + Math.random() * 20, type));
        }
    }

    emitExplosion(x, y, color) {
        // Sparks
        this.emit(x, y, color, 15, 'spark');
        // Shockwave
        this.particles.push(new Particle(x, y, color, 0, 30, 'shockwave'));
    }
    
    emitText(x, y, text, color = '#fff') {
         this.particles.push(new Particle(x, y, color, 0, 40, 'text', text));
    }

    update() {
        for (let i = this.particles.length - 1; i >= 0; i--) {
            let p = this.particles[i];
            p.update();
            if (p.life <= 0) {
                this.particles.splice(i, 1);
            }
        }
    }
}
